lua/lazy-lsp/helpers.lua

```diff
@@ -16,8 +16,76 @@ local function escape_shell_args(args)
   return table.concat(escaped, " ")
 end
 
+local cache_nix_command_available = nil
+-- Check whether both `nix` command and `nixpkgs` flake are available
+---@return boolean?
+local function nix_command_available()
+  if cache_nix_command_available ~= nil then
+    return cache_nix_command_available
+  end
+  cache_nix_command_available = false
+
+  -- Check that both `nix-command` and `flake` options are in effect
+  local registry = vim.fn.system({ "nix", "registry", "list", "--offline" })
+  if vim.v.shell_error == 0 then
+    for flake in vim.gsplit(registry, "\n") do
+      -- Nixpkgs flake is available
+      local flake_url = string.match(flake, "^%S+ flake:nixpkgs (.*)")
+      if flake_url then
+        -- And we won't accidentally fetch `nixpkgs-unstable` as it used to be.
+        -- Only since NixOS 24.05, system has its nixpkgs flake in the registry by default
+        if string.match(flake_url, "^path:") then
+          cache_nix_command_available = true
+        end
+        break
+      end
+    end
+  end
+  return cache_nix_command_available
+end
+
+---@param nix_pkgs string[] | string
+---@param cmd string[]
+---@return string[]?
+local function in_shell_extendable(nix_pkgs, cmd)
+  if type(nix_pkgs) == "string" then
+    nix_pkgs = { nix_pkgs }
+  end
+  if #nix_pkgs == 0 then
+    error("No nix pkg provided")
+    return {}
+  end
+  if not nix_command_available() then
+    return
+  end
+
+  local nix_cmd = { "nix", "shell" }
+  for _, nix_pkg in ipairs(nix_pkgs) do
+    table.insert(nix_cmd, "nixpkgs#" .. nix_pkg)
+  end
+  table.insert(nix_cmd, "--command")
+  vim.list_extend(nix_cmd, cmd)
+  return nix_cmd
+end
+
+---@param nix_pkgs string[] | string
+---@param cmd string[]
+---@return string[]
 local function in_shell(nix_pkgs, cmd)
-  local nix_cmd = { "nix-shell", "-p" }
+  if type(nix_pkgs) == "string" then
+    nix_pkgs = { nix_pkgs }
+  end
+  if #nix_pkgs == 0 then
+    error("No nix pkg provided")
+    return {}
+  end
+
+  local nix_cmd = in_shell_extendable(nix_pkgs, cmd)
```

Comment:
it would be more readable to not have extra function `in_shell_extendable`, it can be removed and code inlined, and have the condition here:
```lua
if not nix_command_available() then
-- code from in_shell_extendable
...
```